parquet_file/src/catalog.rs

```diff
@@ -449,19 +490,29 @@ impl OpenTransaction {
                 .fail()?;
             }
             proto::transaction::action::Action::AddParquet(a) => {
-                if !state.parquet_files.insert(a.path.clone()) {
-                    ParquetFileAlreadyExists {
-                        path: a.path.clone(),
+                let path = parse_dirs_and_filename(&a.path)?;
+                match state.parquet_files.entry(path) {
+                    Occupied(o) => {
+                        return Err(Error::ParquetFileAlreadyExists {
+                            path: o.key().clone(),
+                        });
+                    }
+                    Vacant(v) => {
+                        let metadata = thrift_to_parquet_metadata(&a.metadata)
+                            .context(MetadataDecodingFailed {})?;
```

Comment:
You can leave these curly braces out if you'd like:

```suggestion
                            .context(MetadataDecodingFailed)?;
```